### ISSUE TITLE
Support Promise cancellation

### ISSUE THREAD

**Issue by clue** [reactphp/socket-client#40]
We should register a Promise cancellation handler so that the following code actually cleans up the underlying socket resource:

``` php
$promise = $connector->create('reactphp.org', 80);

$promise->cancel();
```

This can eventually also be used as the basis for timeout support (#28), similar to this:

``` php
$promise = $connector->create('reactphp.org', 80);
$loop->addTimer(3.0, function () use ($promise) {
    $promise->cancel();
});
```

**Comment by clue** [comment]
This also likely depends on https://github.com/reactphp/dns/issues/28 so that the DNS request can also be canceled.

**Comment by clue** [comment]
> This also likely depends on reactphp/dns#28 so that the DNS request can also be canceled.

Available as of v0.4.3.

I'm now looking into this...

**Comment by arnaud-lb** [comment]
Any help needed on this ? :) I need this to properly timeout requests in the http-client

**Comment by clue** [comment]
Will be resolved via #71 :tada:
